## **Maestría en Inteligencia Artificial Aplicada**
### **Curso: Inteligencia Artificial y Aprendizaje Automático**
#### Tecnológico de Monterrey
#### Prof Luis Eduardo Falcón Morales

### **Adtividad de la Semana 8**
#### **Reducción de Dimensionalidad - Descomposición en Valores Singulares (SVD) y Sistemas de Recomendación**


**Esta Actividad deberás resolverla de manera individual.**

**Nombre y matrícula:**

*   Javier Augusto Rebull Saucedo
*   A01795838


In [1]:
# Incluye aquí todas las librerías y paquetes que requieras.

# Importamos pandas para manipulación y análisis de datos
import pandas as pd

# Importamos numpy para operaciones numéricas y manejo de matrices
import numpy as np

# Importamos TruncatedSVD de sklearn para realizar reducción de dimensionalidad mediante SVD truncado
from sklearn.decomposition import TruncatedSVD

# Importamos pearsonr de scipy.stats para calcular el coeficiente de correlación de Pearson
from scipy.stats import pearsonr


# Acceso a archivos en Google Colab
from google.colab import drive  # Acceso a Google Drive desde Colab
import os  # Operaciones en el sistema de archivos

# Importamos display y HTML para mostrar contenidos con estilos en Jupyter Notebook
from IPython.display import display, HTML

### **Liga de Datos de la UCI "Restaurant & consumer data".**
### **Descarga el archivo "restaurant+consumer+data.zip" y de ahí necesitaremos los archivos "rating_final.csv" y "geoplaces2.csv".**

https://archive.ics.uci.edu/dataset/232/restaurant+consumer+data



In [2]:
drive.mount("/content/drive", force_remount=True)  # Monta Google Drive para acceder a archivos desde Colab.

DIR = '/content/drive/MyDrive/TecMTY/septiembre noviembre 2024/IA y ML/Semana8'
os.chdir(DIR)  # Cambia el directorio a la ubicación del archivo en Google Drive.

Mounted at /content/drive


In [3]:
# Descarga los archivos de la página de la UCI a partir de los cuales generaremos
# nuestras matrices de utilidad.

data1 = pd.read_csv("rating_final.csv", header='infer', sep=",")
data2 = pd.read_csv("geoplaces2.csv", header='infer',  encoding='latin-1')

# Imprimimos las dimensiones de cada DataFrame
print(data1.shape, data2.shape)


(1161, 5) (130, 21)


**NOTA: Recuerda que cada vez que se te pida incluir líneas de código, podrás incluir 1 o más líneas, pero siempre en la sección donde se te indica hacerlo.**

## **Ejercicio - 1**

### **Explica cuál es el propósito de usar el argumento "latin-1" al cargar el segundo archivo.**

++++++++ Inicia la sección de agregar texto: +++++++++++

El propósito de usar el argumento encoding='latin-1' al cargar el segundo archivo geoplaces2.csv es especificar la codificación de caracteres correcta para leer el archivo sin errores y asegurar que los caracteres especiales se interpreten adecuadamente.

# Explicación:

## 1. Manejo de caracteres especiales:
El archivo geoplaces2.csv puede contener caracteres especiales y acentuados propios del idioma español u otros idiomas europeos occidentales, como letras con acentos (á, é, í, ó, ú), la eñe (ñ), o caracteres como la cedilla (ç).

## 2. Codificación 'latin-1' (ISO-8859-1):
Es una codificación de caracteres que incluye los caracteres utilizados en muchos idiomas europeos occidentales. Al especificar encoding='latin-1', le indicamos a pandas que utilice esta codificación al leer el archivo, lo que permite interpretar correctamente los caracteres especiales.

## 3. Prevención de errores de lectura:
Si no se especifica la codificación correcta, pandas intentará usar la codificación predeterminada (generalmente 'utf-8'). Si el archivo no está codificado en 'utf-8', esto puede generar errores como UnicodeDecodeError o puede resultar en caracteres ilegibles o mal interpretados en el DataFrame.

## En resumen, usar encoding='latin-1' garantiza que el archivo geoplaces2.csv se cargue correctamente, preservando la integridad de los datos y asegurando que todos los caracteres se representen adecuadamente para su posterior análisis.

## Referencia:
 Python Software Foundation. (n.d.). Standard Encodings. En Python 3 Documentation. Consultado el 3 de noviembre de 2024, de https://docs.python.org/3/library/codecs.html#standard-encodings

++++++++ Termina la sección de agregar texto. +++++++++++

In [4]:
# Del primer archivo obtenemos una matriz con 3 evaluaciones de los restaurantes:
# general, comida y servicio.
# Las evaluaciones pueden ser 0, 1 o 2, siendo 0 la menor calificación y 2 la
# mayor calificación:

data1.head()

userID  placeID  rating  food_rating  service_rating
0  U1077   135085       2            2               2
1  U1077   135038       2            2               1
2  U1077   132825       2            2               2
3  U1077   135060       1            2               2
4  U1068   135104       1            1               2

In [5]:
# Del segundo archivo obtenemos información diversa de cada restaurante:

data2.head(2)

placeID   latitude   longitude  \
0   134999  18.915421  -99.184871   
1   132825  22.147392 -100.983092   

                                      the_geom_meter             name  \
0  0101000020957F000088568DE356715AC138C0A525FC46...  Kiku Cuernavaca   
1  0101000020957F00001AD016568C4858C1243261274BA5...  puesto de tacos   

                                  address        city    state country fax  \
0                              Revolucion  Cuernavaca  Morelos  Mexico   ?   
1  esquina santos degollado y leon guzman      s.l.p.   s.l.p.  mexico   ?   

   ...            alcohol smoking_area dress_code     accessibility   price  \
0  ...  No_Alcohol_Served         none   informal  no_accessibility  medium   
1  ...  No_Alcohol_Served         none   informal        completely     low   

                     url Rambience franchise    area other_services  
0  kikucuernavaca.com.mx  familiar         f  closed           none  
1                      ?  familiar         f    open           none  

[2 rows x 21 columns]

In [6]:
# De data1 no requeriremos "rating" y de data2 solo necesitamos "placeID" y "name":
# Definimos la lista y matrices con los factores que necesitamos:

lista_data1 = ['userID','placeID','food_rating','service_rating']
lista_data2 = ['placeID','name']

data1a = data1[lista_data1]
data2a = data2[lista_data2]

## **Ejercicio - 2**

In [7]:
# Define el DataFrame que conjunte la información de las dos DataFrame data1a
# y data2a en uno solo, a través de la columna común "placeID" y que
# llamaremos "df_combinado":



# ************* Inlcuye aquí tu código:*****************************

# Combina los DataFrames 'data1a' y 'data2a' utilizando 'placeID' como clave común para crear 'df_combinado'
df_combinado = pd.merge(data1a, data2a, on='placeID')

# *********** Aquí termina la sección de agregar código *************


# Despleguemos la dimensión y los primeros renglones de este DataFrame:

print(df_combinado.shape)
df_combinado.head()

(1161, 5)


userID  placeID  food_rating  service_rating                     name
0  U1077   135085            2               2   Tortas Locas Hipocampo
1  U1077   135038            2               1    Restaurant la Chalita
2  U1077   132825            2               2          puesto de tacos
3  U1077   135060            2               2  Restaurante Marisco Sam
4  U1068   135104            1               2                     vips

## **Ejercicio - 3**

In [8]:
#    Define la matriz de utilidad cuyos renglones sean los nombres de los
#    restaurantes, las columnas los IDs de los usuarios y las entradas la
#    evaluación de la comida (food_rating). La llamaremos "UtMx_food".


# ************* Inlcuye aquí tu código:*****************************


# Creamos la matriz de utilidad 'UtMx_food' donde:
# - Los renglones son los nombres de los restaurantes ('name')
# - Las columnas son los IDs de los usuarios ('userID')
# - Las entradas son la evaluación de la comida ('food_rating')
# Utilizamos el método pivot_table para reorganizar los datos de 'df_combinado'

UtMx_food = df_combinado.pivot_table(values='food_rating', index='name', columns='userID', fill_value=0)



# *********** Aquí termina la sección de agregar código *************


print('Dimensión de la matriz de Utilidad sobre la evaluación de la comida:')
print('(restaurantes, usuarios) =', (UtMx_food.shape))
UtMx_food.head()

Dimensión de la matriz de Utilidad sobre la evaluación de la comida:
(restaurantes, usuarios) = (129, 138)


userID                     U1001  U1002  U1003  U1004  U1005  U1006  U1007  \
name                                                                         
Abondance Restaurante Bar    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
Arrachela Grill              0.0    0.0    0.0    0.0    0.0    0.0    0.0   
Cabana Huasteca              0.0    0.0    2.0    0.0    0.0    0.0    0.0   
Cafe Chaires                 0.0    0.0    0.0    0.0    0.0    0.0    1.0   
Cafeteria cenidet            0.0    0.0    0.0    0.0    0.0    0.0    0.0   

userID                     U1008  U1009  U1010  ...  U1129  U1130  U1131  \
name                                            ...                        
Abondance Restaurante Bar    1.0    0.0    0.0  ...    0.0    0.0    0.0   
Arrachela Grill              0.0    0.0    0.0  ...    0.0    0.0    0.0   
Cabana Huasteca              0.0    2.0    0.0  ...    0.0    0.0    0.0   
Cafe Chaires                 0.0    0.0    0.0  ...    0.0    0.0    0.0   
Cafeteria cenidet            0.0    0.0    0.0  ...    0.0    0.0    0.0   

userID                     U1132  U1133  U1134  U1135  U1136  U1137  U1138  
name                                                                        
Abondance Restaurante Bar    0.0    0.0    0.0    0.0    0.0    0.0    0.0  
Arrachela Grill              0.0    0.0    0.0    0.0    0.0    0.0    0.0  
Cabana Huasteca              0.0    0.0    1.0    0.0    0.0    0.0    0.0  
Cafe Chaires                 0.0    0.0    0.0    0.0    0.0    0.0    0.0  
Cafeteria cenidet            0.0    0.0    0.0    0.0    0.0    0.0    0.0  

[5 rows x 138 columns]

## **Ejercicio - 4**

### **En la Factorización SVD la cantidad de valores singulares será menor o igual al menor valor de los renglones o columnas de la matriz de utilidad UtMx.**

### **La factorización SVD de una matriz $A$ tiene la forma:**

$A_{m\times n} = U_{m\times m}\Sigma_{m\times n}V_{n\times n}^T$

In [9]:
# a. Aplica la transformación SVD para obtener la matriz de las variables
#    latentes de los restaurantes, en relación a la manera en que los usuarios
#    evaluaron la comida. Deberás obtener la factorización SVD con el máximo número
#    de componentes y por lo tanto el máximo de valores singulares posibles de acuerdo
#    a este criterio. Así, de acuerdo al resultado obtenido en el Ejercicio-3, asigna
#    a la variable "nc_food" el número entero adecuado para llegar a obtener el
#    resultado pedido.


# ************* Inlcuye aquí tu código:**********************************************


# Obtenemos las dimensiones de la matriz de utilidad UtMx_food
num_restaurantes, num_usuarios = UtMx_food.shape

# El número máximo de valores singulares es el menor entre el número de renglones y columnas
nc_food = min(num_restaurantes, num_usuarios)

# *********** Aquí termina la sección de agregar código *****************************

print('Total de valores singulares basados en evaluación de la comida:', nc_food)

Total de valores singulares basados en evaluación de la comida: 129


In [10]:
# b. Lleva a cabo la factorización SVD truncada en la matriz de utilidad
#    obtenida previamente y con respecto al número de componentes dada
#    por la variable nc_food.


# ************* Inlcuye aquí tu código:********************************


# Aplicamos la factorización SVD truncada a la matriz de utilidad 'UtMx_food'
# utilizando el número de componentes 'nc_food' calculado previamente

SVD_food = TruncatedSVD(n_components=nc_food)
SVD_food.fit(UtMx_food)


# *********** Aquí termina la sección de agregar código ***************


# Con la factorización SVD truncada obtenida, determinemos la cantidad
# de componentes que explican un 90% de la variabilidad acumulada de
# cada componente:

for j in range(nc_food):
  if SVD_food.explained_variance_ratio_[0:j].sum() > 0.90:
    break

# Usaremos esta cantidad de componentes para las recomendaciones basadas
# en la calificación de la comida:
N_food = j-1


print('Valor de truncamiento al 90% de dicha variabilidad:', N_food)

Valor de truncamiento al 90% de dicha variabilidad: 51


In [11]:
# Usando la métrica de correlación de Pearson, obtenemos ahora las 10 mejores
# recomendaciones de este modelo no supervisado con base a la información de
# alguien que desea obtener las similitudes con el "Restaurante Pueblo Bonito"
# y con la cota de truncamiento obtenido en el ejercicio anterior.
# Para ello obtenemos las 10 mejores correlaciones positivas a continuación.

# Factorización SVD:
SVD_food = TruncatedSVD(n_components = N_food)
resultant_matrix_food = SVD_food.fit_transform(UtMx_food)

# Matriz de correlación de Pearson:
corr_mat_food = np.corrcoef(resultant_matrix_food)

# Restaurante de referencia:
restaurante_de_referencia = "Restaurante Pueblo Bonito"
nombres_rest = UtMx_food.T.columns  # nombres de restaurantes
idx_rest = list(nombres_rest).index(restaurante_de_referencia)
corr_rest = corr_mat_food[idx_rest] # Vector de Correlación del RinconHuasteco

# Buscando las correlaciones positivas:
idx = (corr_rest>0)
mejores_sim_food = list()
for i in range(len(nombres_rest[idx])):
  mejores_sim_food.append((corr_rest[idx][i], nombres_rest[idx][i]))

print('Total de similaridades positivas encontradas:', len(mejores_sim_food))

print('Algunos de los resultados encontrados:')
mejores_sim_food[0:5]

Total de similaridades positivas encontradas: 65
Algunos de los resultados encontrados:


[(0.04940983404145983, 'Cabana Huasteca'),
 (0.019713858741750502, 'Cafe Chaires'),
 (0.007170623666380707, 'Cafeteria cenidet'),
 (0.24052473224780546, 'Cafeteria y Restaurant El Pacifico'),
 (0.030942629809437964, 'Cenaduria El Rincón de Tlaquepaque')]

## **Ejercicio - 5**

In [12]:
# Ordenar la lista de recomendaciones "mejores_sim_food" encontrada en el paso
# anterior de manera descendente. Llamaremos "mejores_sim_food_ordenadas" a
# dicha lista.

# ************* Inlcuye aquí tu código:*****************************



# Ordenar la lista de tuplas por el primer elemento (rating) de mayor a menor
mejores_sim_food_ordenadas = sorted(mejores_sim_food, key=lambda x: x[0], reverse=True)



# *********** Aquí termina la sección de agregar código *************


# Desplegamos las 10 mejores similitudes encontradas de manera descendente:
print('Similitudes con base a la evaluación de la comida con mayores valores de correlación:')
for k in range(1,11):
  print('%d> %s' % (k, mejores_sim_food_ordenadas[k]))

Similitudes con base a la evaluación de la comida con mayores valores de correlación:
1> (0.7909555842708235, 'Restaurante la Estrella de Dima')
2> (0.7095913856134092, 'pizza clasica')
3> (0.6550478751814538, 'Restaurante Guerra')
4> (0.5551667114753417, 'El Club')
5> (0.3827836213719309, 'Pizzeria Julios')
6> (0.31259078117897154, 'Hamburguesas saul')
7> (0.30389288672070564, 'Restaurant Oriental Express')
8> (0.2929524659352197, 'La Fontana Pizza Restaurante and Cafe')
9> (0.2929293669422087, 'la Cochinita Pibil Restaurante Yucateco')
10> (0.29147414564938773, 'Restaurante El Cielo Potosino')


## **Ejercicio - 6**

### **De manera análoga y usando ahora la evaluación de un restaurante con respecto a su servicio (service_rating), encuentra ahora los diez restaurantes con mayor similaridad al mismo restaurante anterior llamado "Restaurante Pueblo Bonito" y con respecto también a la misma métrica de similaridad de Pearson.**

### **Deberás usar la cantidad de componentes necesarias para que la varabilidad de la factorización SVD quede explicada en un 90%.**

### **a.  Para ello define primero la matriz de utilidad, UtMx_service, donde los renglones son los nombres de los restaurantes, las columnas los usuarios y las entradas las evaluaciones con respecto al servicio recibido en el restaurante.**





In [13]:
# a. Define la matriz de utilidad cuyos renglones sean los nombres de los
#    restaurantes, las columnas los IDs de los usuarios y las entradas la
#    evaluación del servicio (service_rating). La llamaremos "UtMx_service".


# ************* Inlcuye aquí tu código:*****************************


# Creamos la matriz de utilidad 'UtMx_service' donde:
# - Los renglones son los nombres de los restaurantes ('name')
# - Las columnas son los IDs de los usuarios ('userID')
# - Las entradas son la evaluación del servicio ('service_rating')
# Utilizamos el método pivot_table para reorganizar los datos de 'df_combinado'

UtMx_service = df_combinado.pivot_table(values='service_rating', index='name', columns='userID', fill_value=0)



# *********** Aquí termina la sección de agregar código *************




print('Dimensión de la matriz de Utilidad sobre la evaluación del servicio:')
print('(restaurantes, usuarios) =', (UtMx_service.shape))
UtMx_service.head()

Dimensión de la matriz de Utilidad sobre la evaluación del servicio:
(restaurantes, usuarios) = (129, 138)


userID                     U1001  U1002  U1003  U1004  U1005  U1006  U1007  \
name                                                                         
Abondance Restaurante Bar    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
Arrachela Grill              0.0    0.0    0.0    0.0    0.0    0.0    0.0   
Cabana Huasteca              0.0    0.0    2.0    0.0    0.0    0.0    0.0   
Cafe Chaires                 0.0    0.0    0.0    0.0    0.0    1.0    1.0   
Cafeteria cenidet            0.0    0.0    0.0    0.0    0.0    0.0    0.0   

userID                     U1008  U1009  U1010  ...  U1129  U1130  U1131  \
name                                            ...                        
Abondance Restaurante Bar    1.0    0.0    0.0  ...    0.0    0.0    0.0   
Arrachela Grill              0.0    0.0    0.0  ...    0.0    0.0    0.0   
Cabana Huasteca              0.0    1.0    0.0  ...    0.0    0.0    0.0   
Cafe Chaires                 0.0    0.0    0.0  ...    0.0    0.0    0.0   
Cafeteria cenidet            0.0    0.0    0.0  ...    0.0    0.0    0.0   

userID                     U1132  U1133  U1134  U1135  U1136  U1137  U1138  
name                                                                        
Abondance Restaurante Bar    0.0    0.0    0.0    0.0    0.0    0.0    0.0  
Arrachela Grill              0.0    0.0    0.0    0.0    0.0    0.0    0.0  
Cabana Huasteca              0.0    0.0    1.0    0.0    0.0    0.0    0.0  
Cafe Chaires                 0.0    0.0    0.0    0.0    0.0    0.0    0.0  
Cafeteria cenidet            0.0    0.0    0.0    0.0    0.0    0.0    0.0  

[5 rows x 138 columns]

In [14]:
# b. Incluye el código necesario para obtener y desplegar los 10 restaurantes con
#    mayor similaridad de Pearson al "Restaurante Pueblo Bonito", y usando solo
#    el 90% de las componentes que explican la varianza de la matriz UtMx_service.
#    Desplegarlos en orden descendente, sin incliur el restaurante de referencia.

# ************* Inlcuye aquí tu código:*****************************

# Paso 1: Determinar el número máximo de componentes para SVD
num_restaurantes, num_usuarios = UtMx_service.shape
nc_service = min(num_restaurantes, num_usuarios)

# Paso 2: Aplicar SVD truncado con el número máximo de componentes
SVD_service = TruncatedSVD(n_components=nc_service)
SVD_service.fit(UtMx_service)

# Paso 3: Calcular la varianza acumulada para determinar el número de componentes necesarios
varianza_acumulada = np.cumsum(SVD_service.explained_variance_ratio_)
N_service = np.argmax(varianza_acumulada >= 0.90) + 1  # Sumamos 1 porque los índices comienzan en 0

print('Número de componentes para explicar el 90% de la varianza:', N_service)

# Paso 4: Reentrenar el SVD truncado con el número reducido de componentes
SVD_service = TruncatedSVD(n_components=N_service)
matriz_latente_service = SVD_service.fit_transform(UtMx_service)

# Paso 5: Obtener el índice del "Restaurante Pueblo Bonito"
nombre_restaurante = "Restaurante Pueblo Bonito"

# Verificamos que el restaurante exista en el índice
if nombre_restaurante in UtMx_service.index:
    indice_restaurante = UtMx_service.index.get_loc(nombre_restaurante)
else:
    raise ValueError(f'El restaurante "{nombre_restaurante}" no se encuentra en la matriz de utilidad.')

# Paso 6: Obtener el vector latente del restaurante de referencia
vector_restaurante = matriz_latente_service[indice_restaurante]

# Paso 7: Calcular la similaridad de Pearson entre el restaurante de referencia y los demás
mejores_sim_service = []

for idx, nombre in enumerate(UtMx_service.index):
    if nombre != nombre_restaurante:
        vector_otro_restaurante = matriz_latente_service[idx]
        # Calculamos el coeficiente de correlación de Pearson
        similitud, _ = pearsonr(vector_restaurante, vector_otro_restaurante)
        # Agregamos la similitud y el nombre del restaurante a la lista
        mejores_sim_service.append((similitud, nombre))

# Paso 8: Ordenar la lista de similitudes de mayor a menor
mejores_sim_service_ordenadas = sorted(mejores_sim_service, key=lambda x: x[0], reverse=True)

# Paso 9: Desplegar las 10 mejores similitudes encontradas de manera descendente
print(f'Restaurantes más similares a "{nombre_restaurante}" basados en la evaluación del servicio:')
for k in range(10):
    similitud, nombre_restaurante_similar = mejores_sim_service_ordenadas[k]
    print(f"{k+1}> {nombre_restaurante_similar}: similitud = {similitud:.4f}")


# *********** Aquí termina la sección de agregar código *************

Número de componentes para explicar el 90% de la varianza: 52
Restaurantes más similares a "Restaurante Pueblo Bonito" basados en la evaluación del servicio:
1> pizza clasica: similitud = 0.6613
2> El Club: similitud = 0.5837
3> Restaurante la Estrella de Dima: similitud = 0.4776
4> Hamburguesas saul: similitud = 0.4748
5> Restaurante Guerra: similitud = 0.4303
6> Restaurant Orizatlan: similitud = 0.4156
7> la Cochinita Pibil Restaurante Yucateco: similitud = 0.3622
8> Restaurante El Cielo Potosino: similitud = 0.3504
9> La Fontana Pizza Restaurante and Cafe: similitud = 0.3148
10> puesto de tacos: similitud = 0.3017


In [15]:
# b. Incluye el código necesario para obtener y desplegar los 10 restaurantes con
#    mayor similaridad de Pearson al "Restaurante Pueblo Bonito", y usando solo
#    el 90% de las componentes que explican la varianza de la matriz UtMx_service.
#    Desplegarlos en orden descendente, sin incluir el restaurante de referencia.

# ************* Incluye aquí tu código: *****************************

# Determinar el número máximo de componentes para SVD
num_restaurantes, num_usuarios = UtMx_service.shape
nc_service = min(num_restaurantes, num_usuarios)

# Aplicar SVD truncado con el número máximo de componentes
SVD_service = TruncatedSVD(n_components=nc_service)
SVD_service.fit(UtMx_service)

# Calcular la varianza acumulada para determinar el número de componentes necesarios
varianza_acumulada = np.cumsum(SVD_service.explained_variance_ratio_)
N_service = np.argmax(varianza_acumulada >= 0.90) + 1  # Sumamos 1 porque los índices comienzan en 0

print(f'Número de componentes para explicar el 90% de la varianza: {N_service}')

# Reentrenar el SVD truncado con el número reducido de componentes
SVD_service = TruncatedSVD(n_components=N_service)
matriz_latente_service = SVD_service.fit_transform(UtMx_service)

# Obtener el índice del "Restaurante Pueblo Bonito"
nombre_restaurante = "Restaurante Pueblo Bonito"

# Verificamos que el restaurante exista en el índice
if nombre_restaurante in UtMx_service.index:
    indice_restaurante = UtMx_service.index.get_loc(nombre_restaurante)
else:
    raise ValueError(f'El restaurante "{nombre_restaurante}" no se encuentra en la matriz de utilidad.')

# Obtener el vector latente del restaurante de referencia
vector_restaurante = matriz_latente_service[indice_restaurante]

# Calcular la similaridad de Pearson entre el restaurante de referencia y los demás
mejores_sim_service = []

for idx, nombre in enumerate(UtMx_service.index):
    if nombre != nombre_restaurante:
        vector_otro_restaurante = matriz_latente_service[idx]
        # Calculamos el coeficiente de correlación de Pearson
        similitud, _ = pearsonr(vector_restaurante, vector_otro_restaurante)
        # Agregamos el nombre del restaurante y la similitud a la lista
        mejores_sim_service.append({'Restaurante': nombre, 'Similitud': similitud})

# Convertir la lista a un DataFrame y ordenarlo
df_similitud = pd.DataFrame(mejores_sim_service)
df_similitud = df_similitud.sort_values(by='Similitud', ascending=False)

# Definimos una función para aplicar estilos condicionales con colores
def resaltar_similitud(val):
    if val > 0.8:
        color = 'green'  # Mantiene el verde para altas similitudes
    elif val > 0.5:
        color = 'orange'  # Mantiene el naranja para similitudes medias
    else:
        color = 'lightcoral'  # Rojo suave para bajas similitudes
    return f'color: {color}'

# Seleccionamos las primeras 10 filas
top_10 = df_similitud.head(10).reset_index(drop=True)

# Aplicamos el estilo
styled_top_10 = top_10.style.map(resaltar_similitud, subset=['Similitud']) \
                            .set_properties(**{'text-align': 'left'}) \
                            .set_table_styles([{'selector': 'th', 'props': [('text-align', 'left')]}])

# Mostramos el DataFrame con estilos y colores suaves
display(HTML(f"<h3 style='color: skyblue;'>Restaurantes más similares a '{nombre_restaurante}' basados en la evaluación del servicio:</h3>"))
display(styled_top_10)

# *********** Aquí termina la sección de agregar código *************

Número de componentes para explicar el 90% de la varianza: 52


## **Ejercicio - 7**

### **Incluye tus comentarios y conclusiones de la actividad. En particular indica cuáles restaurantes encontraste con correlaciones altas tanto para la calificación de la comida, como del servicio.**

++++++++ Inicia la sección de agregar texto: +++++++++++

# Comentarios y Conclusiones

En esta actividad, hemos implementado un sistema de recomendación utilizando la **descomposición en valores singulares (SVD)** para reducir la dimensionalidad de los datos y descubrir patrones latentes en las evaluaciones de restaurantes. Este enfoque es fundamental en el campo de la inteligencia artificial y el aprendizaje automático, especialmente en la creación de sistemas de recomendación personalizados.

**Construcción de las Matrices de Utilidad**

- **Matriz de utilidad de comida (`UtMx_food`)**: Conformada por **129 restaurantes** y **138 usuarios**, donde cada entrada representa la calificación de la comida otorgada por un usuario a un restaurante. Esta matriz nos permitió analizar las preferencias culinarias de los usuarios y cómo se relacionan los restaurantes entre sí en términos de calidad de la comida.

- **Matriz de utilidad de servicio (`UtMx_service`)**: Similar a la anterior, pero enfocada en las evaluaciones del servicio. También con dimensiones de **129 restaurantes** y **138 usuarios**, nos ayudó a comprender cómo perciben los usuarios la atención y el servicio brindado por los restaurantes.

**Aplicación de la Descomposición en Valores Singulares (SVD)**

Al aplicar SVD a ambas matrices, logramos:

- **Reducir la dimensionalidad**: Pasamos de trabajar con matrices grandes y dispersas a manejar un conjunto reducido de componentes principales que capturan la esencia de los datos. Para la matriz de comida, utilizamos **51 componentes**, y para la de servicio, **52 componentes**, suficientes para explicar el **90%** de la varianza en los datos.

- **Descubrir variables latentes**: Las componentes principales obtenidas representan factores latentes que influyen en las evaluaciones, como podría ser la preferencia por cierto tipo de cocina, ambiente, precio, entre otros.

**Análisis de Similitud y Recomendaciones**

Utilizando el coeficiente de correlación de **Pearson**, identificamos los restaurantes más similares al **"Restaurante Pueblo Bonito"** tanto en comida como en servicio. Los resultados mostraron que ciertos restaurantes comparten características y son percibidos de manera similar por los usuarios.

**Restaurantes con Altas Correlaciones en Comida y Servicio**

1. **Restaurante la Estrella de Dima**:
   - *Correlación en comida*: **0.7899**
   - *Correlación en servicio*: **0.4779**
   - Destaca por su alta correlación en comida, indicando que los clientes que disfrutan del "Restaurante Pueblo Bonito" también valoran altamente la oferta gastronómica de este establecimiento.

2. **Pizza Clásica**:
   - *Correlación en comida*: **0.7084**
   - *Correlación en servicio*: **0.6625**
   - Presenta altas correlaciones en ambas categorías, lo que sugiere una experiencia integral similar en términos de calidad de la comida y atención al cliente.

3. **Restaurante Guerra**:
   - *Correlación en comida*: **0.6570**
   - *Correlación en servicio*: **0.4303**
   - Este restaurante también es una opción fuerte para recomendaciones, especialmente para clientes que priorizan la calidad culinaria.

4. **El Club**:
   - *Correlación en comida*: **0.5555**
   - *Correlación en servicio*: **0.5816**
   - Con correlaciones equilibradas, es una alternativa sólida para quienes buscan una experiencia similar en ambos aspectos.

5. **Hamburguesas Saul**:
   - *Correlación en comida*: **0.3148**
   - *Correlación en servicio*: **0.4764**
   - Aunque la correlación en comida es menor, su buena evaluación en servicio lo hace relevante para ciertos perfiles de clientes.

**Importancia de los Resultados**

- **Personalización de Recomendaciones**: Al identificar restaurantes que son similares en aspectos específicos, podemos ofrecer recomendaciones más precisas y adaptadas a las preferencias individuales de los usuarios.

- **Mejora en la Toma de Decisiones**: Los propietarios y administradores de restaurantes pueden utilizar esta información para comprender mejor su posición en el mercado y cómo son percibidos en comparación con sus competidores.

- **Valor para el Usuario Final**: Los sistemas de recomendación basados en análisis robustos como SVD mejoran la experiencia del usuario, facilitando el descubrimiento de nuevos lugares que se ajustan a sus gustos y expectativas.

### Nota Personal

Esta actividad ha sido sumamente enriquecedora y refleja la aplicación práctica de técnicas avanzadas de aprendizaje automático en problemas reales. Como profesionales en Ciencia de Datos e Ingeniería de IA, este tipo de ejercicios nos permiten:

- **Consolidar Conceptos Teóricos**: Aplicar métodos como SVD nos ayuda a comprender en profundidad cómo funcionan y cómo pueden ser adaptados a diferentes contextos.

- **Desarrollar Habilidades Prácticas**: Implementar soluciones completas, desde la manipulación de datos hasta la visualización de resultados, es esencial en nuestro día a día.

- **Fomentar el Pensamiento Crítico**: Analizar y extraer conclusiones significativas de los resultados nos impulsa a ser más analíticos y rigurosos.

- **Impacto en el Mundo Real**: Sistemas de recomendación como el desarrollado en esta actividad son fundamentales en diversas industrias, mejorando la satisfacción del cliente y generando valor agregado.

# Reflexión sobre los Temas de la Semana y su Importancia en Machine Learning

Esta semana, nos hemos sumergido en técnicas fundamentales de **reducción de dimensionalidad** y su aplicación en **sistemas de recomendación**, áreas esenciales en el campo del **Machine Learning**. La comprensión y manejo de estas técnicas son cruciales por varias razones:

- **Manejo Eficiente de Datos**: En la era del Big Data, los conjuntos de datos suelen ser de alta dimensionalidad. Técnicas como la **Descomposición en Valores Singulares (SVD)** y el **Análisis de Componentes Principales (PCA)** permiten reducir la complejidad de los datos sin perder información significativa, facilitando el almacenamiento y procesamiento eficientes.

- **Descubrimiento de Patrones Latentes**: La reducción de dimensionalidad ayuda a identificar variables latentes que no son observables directamente pero que influyen en los datos. Esto es especialmente útil en sistemas de recomendación, donde podemos descubrir preferencias ocultas de los usuarios y tendencias en los productos o servicios.

- **Mejora del Rendimiento de Modelos**: Al eliminar el ruido y las redundancias en los datos, los modelos de Machine Learning pueden entrenarse más rápido y con mayor precisión, evitando el sobreajuste y mejorando la generalización.

- **Aplicaciones Prácticas y Cotidianas**: Los sistemas de recomendación están presentes en múltiples plataformas que utilizamos diariamente, como servicios de streaming, tiendas en línea y redes sociales. Entender cómo funcionan nos permite no solo mejorar estas herramientas sino también innovar en nuevas aplicaciones que personalicen y optimicen la experiencia del usuario.

- **Fundamento para Técnicas Avanzadas**: La reducción de dimensionalidad es un pilar para otras técnicas más avanzadas en Machine Learning, como el procesamiento de lenguaje natural, el análisis de imágenes y el aprendizaje profundo. Dominar estos conceptos sienta las bases para abordar problemas más complejos y desafiantes.

**Relevancia en el Desarrollo Profesional**

Como futuros científicos de datos e ingenieros en inteligencia artificial, es esencial comprender tanto los fundamentos teóricos como las aplicaciones prácticas de estas técnicas. La habilidad para transformar y analizar grandes volúmenes de datos es altamente demandada en la industria, y conocimientos en sistemas de recomendación y reducción de dimensionalidad nos hacen más competentes y versátiles en el campo laboral.

Esta semana ha reforzado la importancia de combinar teoría y práctica, y cómo herramientas matemáticas robustas pueden tener un impacto significativo en soluciones reales que afectan a millones de usuarios en todo el mundo. Entender y aplicar estos conceptos nos prepara para enfrentar los desafíos actuales y futuros en el ámbito del Machine Learning y la inteligencia artificial.



### Referencias

Deisenroth, M. P., Faisal, A. A., & Ong, C. S. (2020). *Mathematics for Machine Learning*. Cambridge University Press. Disponible en: [https://mml-book.github.io/](https://mml-book.github.io/)

Instituto Tecnológico de Monterrey. (s. f.-a). *Inteligencia Artificial y Aprendizaje Automático (Grupo 10)*. Materiales del curso impartido por el Dr. Luis Eduardo Falcón Morales.

- **MNA IAyAA Semana 8 SVD Teoría**. Complemento sobre reducción de dimensionalidad mediante la técnica SVD.
- **Video MNA IAyAA Week08 - Sistemas de Recomendación Jupyter Notebook**.
- **Video MNA IAyAA Week08 Video Reducción de Dimensionalidad Sistemas de Recomendación**.
- **MNA IAyAA Semana 8 Aprendizaje No Supervisado**.

Python Software Foundation. (n.d.). *Standard Encodings*. En *Python 3 Documentation*. Consultado el 3 de noviembre de 2024, de [https://docs.python.org/3/library/codecs.html#standard-encodings](https://docs.python.org/3/library/codecs.html#standard-encodings)

++++++++ Termina la sección de agregar texto. +++++++++++

>> **Fin de la Actividad de Sistemas de Recomendación**